### PR TITLE
Added spectrum whitening

### PR DESCRIPTION
**Pull request by KAdamek**
---
name: Spectrum whitening
about: Spectrum whitening on the GPU added.
The algorithm used is the algorithm used in SIGPROC. The spectrum is divided into disjoint chunks of 128 samples. In each chunk, the mean and standard deviation is calculated with outlier rejection and then the chunk is normalized to mean zero and standard deviation of 1.

---

**Description of pull request**

**Interfaces**
Will this pull request result in a backwards-incompatible interface change: No

Expected semantic version number increment category (Please indicate x.y.z): y

**Issues this pull request solves**

**New tag of master**

**New deployment**

**Keep branch after merging** No

**Notes**

### CODE REVIEW

File: src/aa_device_periods.cu [hunk 1]
```diff
@@ -799,6 +800,15 @@ namespace astroaccelerate {
     (*compute_time) = (*compute_time) + timer.Elapsed();
     //---------<
 	
+	//---------> Spectrum whitening
+	timer.Start();
+	cudaStream_t stream; stream = NULL;
+	spectrum_whitening_SGP1(d_frequency_power, (t_nTimesamples>>1), t_nDMs_per_batch, stream);
```
Comment (jan2nov) [inline src/aa_device_periods.cu:806]: Just to be sure, the acknowledge here is to have the streams at some point and thus we are creating streams as default NULL?

File: src/aa_device_spectral_whitening_kernel.cu [hunk 2]
```diff
@@ -0,0 +1,214 @@
...
+	__shared__ float s_par_MSD[256];
+	__shared__ int s_par_nElements[128];
+
+	float mean, stdev;
+	unsigned long int gpos = blockIdx.y*nSamples;
+	unsigned long int spos = blockIdx.x*blockDim.x;
+	int x_steps = 1;
+	
+	MSD_block(&d_input[gpos + spos], &x_steps, &nSamples, s_par_MSD, s_par_nElements, &mean, &stdev);
+
```
Comment (jan2nov) [inline src/aa_device_spectral_whitening_kernel.cu:194]: Just wondering is it still in the memory size of d_input (mean the gpos + spos)
	dim3 grid_size((nSamples + 128 - 1)/128, nDMs , 1);
	dim3 block_size(128, 1, 1);

**Review by jan2nov — COMMENTED** [review]
Please read all the comments to clarify.

File: src/aa_device_periods.cu (hunk 1 above)
Comment (KAdamek) [reply src/aa_device_periods.cu:806]: Yes, this is for the possibility of including streams in the future. For the kernel to be launched in the default stream we can use NULL.

File: src/aa_device_spectral_whitening_kernel.cu (hunk 2 above)
Comment (KAdamek) [reply src/aa_device_spectral_whitening_kernel.cu:194]: This should be fine. gpos moves the coordinate to the start of the DM-trial while spos is moving coordinates within the DM-trial so gpos is y-axis and spos is x-axis. Besides there is an if branch within the MSD_block function.

File: src/aa_device_periods.cu (hunk 1 above)
Comment (jan2nov) [reply src/aa_device_periods.cu:806]: OK

File: src/aa_device_spectral_whitening_kernel.cu (hunk 2 above)
Comment (jan2nov) [reply src/aa_device_spectral_whitening_kernel.cu:194]: I missed the if branch probably.

**Pull request merged**